Yaml template finalization

Ancairon opened the pull request:
##### Summary

Let's discuss on this template, so then I can generate all of netdata/netdata collectors in the same form.

andrewm4894 commented on collectors/python.d.plugin/beanstalk/metadata.yaml:
```diff
@@ -0,0 +1,207 @@
+name: beanstalk
```
will there ever be need for a `name` and an `id` or should `name` be enough to act like an id?

Ancairon replied:
I think name will act as id, so for proc, name would be the individual modules

```
name: /proc/stat
...
name: /proc/sys/kernel/random/entropy_avail
```
that will help us then construct the name, this name is how it comes right out of the CSV files.

andrewm4894 replied:
this is exactly what i'm wondering - how would that then link to blog posts? 

`/proc/sys/kernel/random/entropy_avail`

needs to end up being able to be tagged in blog and linked something like:

https://blog.netdata.cloud/tags/entropy/

so we need, i think, some sort of guid (ideally human readable) that will be a solid key (not special chars, just text if possible) across all places.

so thinking how can we be as flexible as possible here.

Ancairon replied:
Hmm, I see, if it appears in tags, it should probably not be proc.plugin.entropy_avail..

From the script I am generating this I don't really have much to work with, I have the columns the CSV has.

So if we want something like "entropy", "statistics" and stuff, we might need to manually add it upon filling in the yamls :thinking:, what do you think?

Another idea would be to make the tags in the blogs like the last /../../../X element in the path, or even leave the tag as "entropy" and then search with regex if it exists in that last module bit :thinking:

Ancairon commented on collectors/python.d.plugin/beanstalk/metadata.yaml:
```diff
@@ -0,0 +1,207 @@
+  supported_platforms:
+    description: |
+      If this collector can't run on some platforms specify them here. Leave empty otherwise.
+  multi-instance: true
```
Multi instance is whether or not this supports multiple Jobs, essentially what the [template](https://docs.google.com/document/d/19_vYbp89AkMA_Is4tjnYd9p-UuhALVwS4eeH-mpGfPg/edit?usp=sharing) indicates

>Does it support multiple jobs?
Example: The collector supports collecting metrics from multiple apache servers, local or remote to the node the collector runs.

We will say / no, and have templated messages in the CI, I believe this is cleanest

sashwathn replied:
@Ancairon / @ilyam8 : Are there any collectors that do not support multiple instances?

ilyam8 replied:
Yes, all that collect the host metrics are a single instance.

andrewm4894 commented on collectors/python.d.plugin/beanstalk/metadata.yaml:
```diff
@@ -0,0 +1,207 @@
+alerts:
+  - name: beanstalk_server_buried_jobs
+    link: https://gihtub.com/netdata/netdata/blob/master/health/health.d/beanstalkd.conf
+    metric: beanstalk.current_jobs
```
i dont think `beanstalk.current_jobs` is a metric - i think its a chart. This alarm seems to be operating on the `buried` dimension of the `beanstalk.current_jobs` chart.

so am wondering what we mean by `metric: beanstalk.current_jobs` ?

andrewm4894 replied:
maybe we mean that the metric is something like `beanstalk.current_jobs.buried` but not sure we really have any such standard representation like this strictly speaking.

Ancairon replied:
That comes from https://github.com/netdata/netdata/blob/master/health/health.d/beanstalkd.conf#L4, it is autogenerated, by parsing the .conf files

Ancairon replied:
Keep in mind this is the template, we can always not render stuff at the final presentation, for example the `os` field in alerts is just for manually curating some system overview modules that end up having metrics but not the alerts in different OSes. That field will not make the end presentation, it will be there so we can remove any wrong alerts.

Same can happen with `metric` field, it is getting whatever the .conf files say

Ancairon commented on collectors/python.d.plugin/beanstalk/metadata.yaml:
```diff
@@ -0,0 +1,207 @@
+        What info I can provide for other plugins that reference me? this will only be visible in their end, they will scrape this text. Example -> XYZ allows monitoring X Y Z metrics...
+alerts:
+  - name: beanstalk_server_buried_jobs
+    link: https://gihtub.com/netdata/netdata/blob/master/health/health.d/beanstalkd.conf
```
nice catch, fixing in the script

andrewm4894 commented on collectors/python.d.plugin/beanstalk/metadata.yaml:
```diff
@@ -0,0 +1,207 @@
+      name: global
+      description: TBD
+      labels: []
+      metrics:
```
use of `metrics` here is potentially new to netdata and maybe a little inconsistent. am wondering if we have to go with contexts > dimensions like in here:

https://london.my-netdata.io/api/v2/contexts
https://london.my-netdata.io/api/v1/charts

andrewm4894 replied:
might just be me but i think of `beanstalk.cpu_usage.user` and `beanstalk.cpu_usage.system` as 2 metrics but this implies they are both under the `beanstalk.cpu_usage` metric

Ancairon replied:
yeah, reason it is metrics, is because we named the CSVs metrics.csv 🤔

Ancairon replied:
Same as above though, this doesn't have to be the final name in the presentation, it is just for the editor's (and maybe the CI's) eyes..

sashwathn replied:
@Ancairon : I think these are contexts in the real essence.. Good to check with @ilyam8  if we want to rename this.

ilyam8 replied:
These are metrics

sashwathn replied:
I agree.. I think the users are not aware of contexts etc which is more a Netdata terminology and we should still call them metrics for a user visible yaml.

Ancairon commented:
@andrewm4894 @ilyam8 I added the schemas as we discussed, had to make "scope" an array, so renamed it to "scopes".

ilyam8 commented on collectors/schema.yaml:
```diff
@@ -0,0 +1,210 @@
+    type: "string"
+  title:
+    type: "string"
+  duplicate_for_virtual_integrations:
```
let's just call it `virtual_integrations` or `aliases`

ilyam8 commented on collectors/schema.yaml:
```diff
@@ -0,0 +1,210 @@
+  overview:
+    type: "object"
+    properties:
+      how_the_collector_works:
```
I think we can call it `data_collection` because it describes how the data is collected (and probably briefly what data is collected).

ilyam8 commented on collectors/schema.yaml:
```diff
@@ -0,0 +1,210 @@
+        properties:
+          description:
+            type: "string"
+      supported_platforms:
+        type: "object"
+        properties:
+          description:
+            type: "string"
```
I think this can be an array

ilyam8 commented on collectors/schema.yaml:
```diff
@@ -0,0 +1,210 @@
+                  type: "string"
+                text:
+                  type: "string"
+      behavior:
```
I think 
- this should be `default_behavior`.
- and be a part of the Overview. Why do you think it is Setup?

Ancairon replied:
Agree for name,

The doc had it there, so I went with that

ilyam8 replied:
Ok, but what are your thoughts? Is default behavior Setup?

Ancairon replied:
Actually I agree that this is overview, I'm with you on this, there is no configuration for it, it just gives out info

Ancairon replied:
I also see that configuration is outside of setup, I will move that under setup too

andrewm4894 commented:
@Ancairon maybe these files should live under a folder if possible to be cleaner eg `collectors/metadata/...`

andrewm4894 commented on collectors/schema.yaml:
```diff
@@ -0,0 +1,212 @@
+    type: "string"
+  title:
+    type: "string"
+  virtual_integrations:
```
can we add comments somewhere to say what a `virtual_integration` is?

andrewm4894 commented on collectors/template.yaml:
```diff
@@ -0,0 +1,88 @@
+name: Collector name
+title: TBD Monitoring
+virtual_integrations:
+  list:
```
what is this?

sashwathn replied:
This is for integrations we want to expose which do not have a dedicated collector.
Eg: Ubuntu, Redhat, Other Linux Distros will create a page which is mostly a copy of the OS Metrics (or some page) and only the description changes..
Eg 2: AWS RDS PostgreSQL and other SaaS Postgres offerings

andrewm4894 commented on collectors/schema.yaml:
```diff
@@ -0,0 +1,212 @@
+        properties:
+          description:
+            type: "string"
+      supported_platforms:
```
should/cloud this be an actual explicit list of allowed `supported_platforms`?

andrewm4894 commented on collectors/multi_module_template.yaml:
```diff
@@ -0,0 +1,91 @@
+    related_resources:
+      integrations:
+        list:
+          [apache, weblog, httpcheck, etc]
```
```suggestion
          [collector_a, collector_b, etc]
```

andrewm4894 reviewed (approved):
makes general sense to me, i think we should try get this over the line and the improve iterate it as we start using it. 

will take too long to merge if we try have it perfect from the start.

andrewm4894 reviewed (changes requested):
i do actually think one change i'd request would be to have these files live in /collectors/metadata/ if possible.

Ancairon commented:
>i do actually think one change i'd request would be to have these files live in /collectors/metadata/ if possible.

@andrewm4894 after discussion with Ilya, we figured meta_template might be better to not cause confusion with metadata.yaml files.

ralphm reviewed (commented):
- Can you expand on how these two files are to be used, particularly the part that details metrics?
- I see a schema.json and a template YAML file, but the template is pretty sparse and doesn't include any properties inside of `scopes`.
- Is there a concrete example of a filled in template? Even a partial one that just lists a few metrics/scopes would be very helpful.
- Where is the metrics section auto-generated from (as stated in the schema)? I thought this file was supposed to be the source of truth, even if we initially generated it from `metrics.csv`?
- Why are labels (only?) at the scope level, and not per metric?
- Shouldn't we have a way to document individual dimensions? I think that instead of an array of strings, dimensions should be represented as a list of objects, which the dimension name in a `name` property.
- `text` is very generic. Can we pick something more specific? E.g. `description`?

- If this file is meant to be the source of truth then I am missing some additional meta data on metrics:
  - `metrics.csv` has a chart type column that is missing here.
  - In my [tools](https://github.com/netdata/ucum) I generate a property on metrics being _additive_ or not. Non-additive metrics cannot be summed. Besides metrics with units of temperature, noise/sound/acidity level, and percentages, this also includes metrics that are entirely independent. An example of the latter is a unique count of something.
- If we are also going to use this as a source document for what is now in `contexts.js`, we need properties that signal that a metric is to be rendered in the summary section, with what chart type, and maybe an explicit index number for ordering. I'm sure there is more and we could do this later, but let's make sure that this document's structure can accomodate that.

    The work I did for the [Feed Ontology](https://github.com/netdata/feed-ontology/) builds on a schema and tools by Elastic for ECS, but we added additional fields for our own purposes.

ilyam8 commented on collectors/meta_template/schema.json:
```diff
@@ -0,0 +1,387 @@
+                    "unit": {
+                      "type": "string"
+                    },
+                    "dimensions": {
+                      "type": "array",
+                      "items": {
+                        "type": "string"
```
item type is object see [apache meta](https://github.com/netdata/go.d.plugin/blob/b2b7637c59ad51f98eaec93285890fb6d7b76900/modules/apache/metadata.yaml#L226-L227)

Ancairon replied:
I had changed this due to not fitting yesterday's yaml schema validator, will change it back :+1:

ralphm commented on collectors/meta_template/schema.json:
```diff
@@ -0,0 +1,387 @@
+                      "type": "string"
+                    },
+                    "unit": {
+                      "type": "string"
```
Might be good to point out here (in a `description`) that the format of units is supposed to follow UCUM. We can also do this a bit later, but as soon as this lands, and the UI changes support them, I'll make sure that all metadata files will have the properly formatted units.

Ancairon replied:
I see this as a task that will take a lot more time for the first iteration, you would have to read through every single metric for every single module and fix the units..

Depends on how many resources we get for this to do it.

cc @sashwathn

ilyam8 commented on collectors/meta_template/schema.json:
```diff
@@ -0,0 +1,393 @@
+                "list": {
+                  "type": "array",
+                  "description": "Each of the list's items represents a row in the available config options table",
+                  "items": {
+                    "type": "object",
+                    "properties": {
+                      "name": {
+                        "type": "string",
+                        "description": "Option name"
+                      },
+                      "description": {
+                        "type": "string",
+                        "description": "Option description"
+                      },
+                      "default_value": {
+                        "type": [
+                          "string",
+                          "number"
+                        ],
+                        "description": "Default value of the option, leave empty if none"
+                      },
+                      "required": {
+                        "type": "boolean",
+                        "description": "A boolean value indicating if this options is required or not. Use true / false"
+                      }
```
Option "details" is missing see [an example](https://github.com/netdata/go.d.plugin/blob/b2b7637c59ad51f98eaec93285890fb6d7b76900/modules/openvpn/metadata.yaml#L49-L62)

Ancairon replied:
Thanks, didn't have this field in my template, updated the schema.

Ancairon commented:
>Can you expand on how these two files are to be used, particularly the part that details metrics?

There will be a script that will generate the metrics and alerts fields for us. Then it will be a matter of copying the rest of the template, pasting it and filling in the info. The schema will be used to keep everything uniform

>I see a schema.json and a template YAML file, but the template is pretty sparse and doesn't include any properties inside of scopes.

I can update the template to include those things, but scope names would require dummy names to illustrate, and we have decided no dummy strings in the template..

>Is there a concrete example of a filled in template? Even a partial one that just lists a few metrics/scopes would be very helpful.

i think Ilya answered, so skipping

>Where is the metrics section auto-generated from (as stated in the schema)? I thought this file was supposed to be the source of truth, even if we initially generated it from metrics.csv?

Noting down how it would look would require sample names for scopes and things..

>Why are labels (only?) at the scope level, and not per metric?

I can see a pretty big Slack thread for this, so skipping too..

>Shouldn't we have a way to document individual dimensions? I think that instead of an array of strings, dimensions should be represented as a list of objects, which the dimension name in a name property.

You are right, Ilya caught this too, I changed the schema to reflect this.

>text is very generic. Can we pick something more specific? E.g. description?

:heavy_check_mark: 

>If this file is meant to be the source of truth then I am missing some additional meta data on >metrics:
>
> - metrics.csv has a chart type column that is missing here.

@ilyam8 should we include all the values that the CSV has in the schema and hence in the generated thing (so the script would add them too)

> - In my [tools](https://github.com/netdata/ucum) I generate a property on metrics being additive or not. Non-additive metrics cannot be summed. Besides metrics with units of temperature, noise/sound/acidity level, and percentages, this also includes metrics that are entirely independent. An example of the latter is a unique count of something.
If we are also going to use this as a source document for what is now in contexts.js, we need properties that signal that a metric is to be rendered in the summary section, with what chart type, and maybe an explicit index number for ordering. I'm sure there is more and we could do this later, but let's make sure that this document's structure can accomodate that.
>The work I did for the [Feed Ontology](https://github.com/netdata/feed-ontology/) builds on a schema and tools by Elastic for ECS, but we added additional fields for our own purposes.

I can understand that these documents (or the behemoth of a document that they will contribute to) will need to be expandable, for now I guess we need to ensure that certain fields as you say allow for expansions/updates. The metrics item is an object, so we can add more stuff in it if we want. Also in this stage, if something is not in the metrics.csv there is not much I can do, for example we tried adding alerts to this, and the result will be needed to be proofed by a human eye, due to us naming metrics from different plugins with the same name (system overview metrics, proc, freebsd and mac have the same metric names but not the same alerts, so we need the "os" property to then remove false alerts from these modules)

ralphm commented:
> [..]
> > I see a schema.json and a template YAML file, but the template is pretty sparse and doesn't include any properties inside of scopes.
> 
> I can update the template to include those things, but scope names would require dummy names to illustrate, and we have decided no dummy strings in the template..
> 
> > Is there a concrete example of a filled in template? Even a partial one that just lists a few metrics/scopes would be very helpful.
> 
> i think Ilya answered, so skipping

I meant with this latest version of the schema, but sure, those examples should do.

> > Where is the metrics section auto-generated from (as stated in the schema)? I thought this file was supposed to be the source of truth, even if we initially generated it from metrics.csv?
> 
> Noting down how it would look would require sample names for scopes and things..

I mean that the schema says that the metrics section is auto-generated. We only do this once from `metrics.csv`, I hope. After this, this file becomes the source of truth and we do away with `metrics.csv` entirely, right?
 
> > Why are labels (only?) at the scope level, and not per metric?
> 
> I can see a pretty big Slack thread for this, so skipping too..

Yep, @ilyam8 made the case that _most_ metrics will have scopes and having labels at that level makes sense there. As long as we can extend the format to include labels at the metric level later, I am fine, and @ilyam8 confirmed we could.

> [..]
> > If this file is meant to be the source of truth then I am missing some additional meta data on >metrics:
> > 
> > * metrics.csv has a chart type column that is missing here.
> 
> @ilyam8 should we include all the values that the CSV has in the schema and hence in the generated thing (so the script would add them too)

Even if you are not going to render them to READMEs or other documentation, having this in the file would help using this as a single source of truth, so we can also generate UI from this.
 
> > * In my [tools](https://github.com/netdata/ucum) I generate a property on metrics being additive or not. Non-additive metrics cannot be summed. Besides metrics with units of temperature, noise/sound/acidity level, and percentages, this also includes metrics that are entirely independent. An example of the latter is a unique count of something.
> >   If we are also going to use this as a source document for what is now in contexts.js, we need properties that signal that a metric is to be rendered in the summary section, with what chart type, and maybe an explicit index number for ordering. I'm sure there is more and we could do this later, but let's make sure that this document's structure can accomodate that.
> >   The work I did for the [Feed Ontology](https://github.com/netdata/feed-ontology/) builds on a schema and tools by Elastic for ECS, but we added additional fields for our own purposes.
> 
> I can understand that these documents (or the behemoth of a document that they will contribute to) will need to be expandable, for now I guess we need to ensure that certain fields as you say allow for expansions/updates. The metrics item is an object, so we can add more stuff in it if we want. Also in this stage, if something is not in the metrics.csv there is not much I can do, for example we tried adding alerts to this, and the result will be needed to be proofed by a human eye, due to us naming metrics from different plugins with the same name (system overview metrics, proc, freebsd and mac have the same metric names but not the same alerts, so we need the "os" property to then remove false alerts from these modules)

Sure, this property not being in metrics.csv right now is fine, I can modify my tooling to add work with the `metadata.yaml` instead, and at it there.

As for the OS specific stuff, I think we need to change this. When you have a heterogeneous environment with metrics from different operating systems using the same context name, the data should have the same semantics for all of them. You need to be able to aggregate them across operating systems, and so on. If this cannot be guaranteed, e.g. because the semantics differ per platform, then they must use a different name.

Both the Open Telemetry and Prometheus projects are quite clear on this, with the former explicitly calling out [OS-specific system metrics](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/system-metrics/#systemos---os-specific-system-metrics) requiring a prefix:

- Prometheus: “As a rule of thumb, aggregations over all the attributes of a given metric SHOULD be meaningful,” [Ref: 🔗](https://prometheus.io/docs/practices/naming/#metric-names).
- OpenTelemetry:

  “Semantic ambiguity SHOULD be avoided. Use prefixed metric names in cases where similar metrics have significantly different implementations across the breadth of all existing metrics. For example, every garbage collected runtime has slightly different strategies and measures. Using a single set of metric names for GC, not divided by the runtime, could create dissimilar comparisons and confusion for end users. (For example, prefer `process.runtime.java.gc*` over `process.runtime.gc.*`.) Measures of many operating system metrics are similarly ambiguous.” [Ref: 🔗](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/#general-guidelines)
  
I think this metadata work doesn't really have that kind of change in scope, but rather as part of the [work](https://github.com/netdata/netdata/issues/13347) we'll do for Agent 2.0.

I think it is fine to not include alert definitions in this metadata, but rather have a separate YAML format and file that replaces our current custom alert definitions. I would even be fine with the metrics definitions living in a separate document (e.g. `metrics.yaml`) instead of this one, and then combining them in whatever scripts we build to generate documentation and UI.

Ancairon commented:
@sashwathn @ilyam8 @ralphm  

I have added to the schema :point_right: `name`, `description`, `unit`, `chart_type` and `dimensions`.

The `label`, `plugin` and `module` columns are inside the schema but in other places, should I add them too to each metric (it will create some noise, but we will not lose the metrics.csv info for each row)

ralphm commented:
@Ancairon no need to duplicate this information. These files are meant to be per module/plugin, right, so all metrics would have the same value.

Similarly, if all metics in a scope have the same set of labels, then they don't need to appear on the metric. If there are cases where a metric does have additional labels, we should introduce a property at the metric level, too. However, @ilyam8 said we don't have such cases right now, so we can do that at a later time.

Ancairon commented:
Then I think we are good to merge this, if everyone approves. Next step would be to make a quick PR after this for a multi-module schema and template that would just use the ones we make here, just in a list fashion. Then I should modify the script for autogeneration to meet the schema we end up with.

Ancairon commented:
please wait a bit, I am preparing a final commit for this, and then you can approve

thiagoftsm commented on collectors/meta_template/schema.json:
```diff
@@ -0,0 +1,462 @@
+  "properties": {
+    "plugin_name": {
+      "type": "string",
+      "description": "Plugin name (e.g. apps.plugin, proc.plugin, go.d.plugin). It must match the name of the binary in the plugins.d directory."
```
We do not have binaries neither  for `charts.d` nor for `python`:

```sh
bash-5.2$ file collectors/charts.d.plugin/charts.d.plugin
collectors/charts.d.plugin/charts.d.plugin: Bourne-Again shell script, ASCII text executable
bash-5.2$ file collectors/python.d.plugin/python.d.plugin
collectors/python.d.plugin/python.d.plugin: Bourne-Again shell script, ASCII text executable
```

Should we not use `executable` instead binaries?

thiagoftsm replied:
@Ancairon this is the unique disagreement  I have, except for this, everything looks good. :)
